validate/error.go

```diff
@@ -13,22 +14,81 @@ const referenceTemplate = "https://github.com/opencontainers/runtime-spec/blob/v
+	ConfigFileExistence
+	// ArtifactsInSignleDir represents the error code of artifacts place test.
+	ArtifactsInSingleDir
+
+	// SpecVersion represents the error code of specfication version test.
+	SpecVersion
+
+	// RootOnNonHyperV represents the error code of root setting test on non hyper-v containers
+	RootOnNonHyperV
+	// RootOnNonHyperV represents the error code of root setting test on hyper-v containers
+	RootOnHyperV
+	// PathFormatOnwindows represents the error code of the path format test on Window
+	PathFormatOnWindows
+	// PathName represents the error code of the path name test
+	PathName
+	// PathExistence represents the error code of the path existence test
+	PathExistence
+	//  ReadonlyFilesystem represents the error code of readonly test
+	ReadonlyFilesystem
+	// ReadonlyOnWindows represents the error code of readonly setting test on Windows
+	ReadonlyOnWindows
+
 	// DefaultFilesystems represents the error code of default filesystems test.
-	DefaultFilesystems ErrorCode = iota
+	DefaultFilesystems
 )
 
 type errorTemplate struct {
 	Level     rfc2119.Level
 	Reference func(version string) (reference string, err error)
 }
 
+var (
+	containerFormatRef = func(version string) (reference string, err error) {
+		return fmt.Sprintf(referenceTemplate, version, "bundle.md#container-format"), nil
+	}
+	specVersionRef = func(version string) (reference string, err error) {
+		return fmt.Sprintf(referenceTemplate, version, "config.md#specification-version"), nil
+	}
+	rootRef = func(version string) (reference string, err error) {
+		return fmt.Sprintf(referenceTemplate, version, "config.md#root"), nil
+	}
+	defaultFSRef = func(version string) (reference string, err error) {
+		return fmt.Sprintf(referenceTemplate, version, "config-linux.md#default-filesystems"), nil
+	}
+)
+
 var ociErrors = map[ErrorCode]errorTemplate{
-	DefaultFilesystems: errorTemplate{
-		Level: rfc2119.Should,
-		Reference: func(version string) (reference string, err error) {
-			return fmt.Sprintf(referenceTemplate, version, "config-linux.md#default-filesystems"), nil
-		},
-	},
+	// NonRFCError represents that an error is not a rfc2119 error
+	// Bundle.md
+	// Container Format
+	ConfigFileExistence:  errorTemplate{Level: rfc2119.Must, Reference: containerFormatRef},
+	ArtifactsInSingleDir: errorTemplate{Level: rfc2119.Must, Reference: containerFormatRef},
+
+	// Config.md
+	// Specification Version
+	SpecVersion: errorTemplate{Level: rfc2119.Must, Reference: specVersionRef},
+	// Root
+	RootOnNonHyperV: errorTemplate{Level: rfc2119.Required, Reference: rootRef},
+	RootOnHyperV:    errorTemplate{Level: rfc2119.Must, Reference: rootRef},
+	// TODO
+	PathFormatOnWindows: errorTemplate{Level: rfc2119.Must, Reference: rootRef},
+	PathName:            errorTemplate{Level: rfc2119.Should, Reference: rootRef},
+	PathExistence:       errorTemplate{Level: rfc2119.Must, Reference: rootRef},
+	ReadonlyFilesystem:  errorTemplate{Level: rfc2119.Must, Reference: rootRef},
```

Comment:
I think you intend this for the spec's:

> * **`readonly`** (bool, OPTIONAL) If true then the root filesystem MUST be read-only inside the container, defaults to false.

But that is a runtime requirement which is tested by `validation` and `runtimetest`.  It doesn't belong in `validate`, which is about config/bundle validation.

Reply:
Yes, it is same with  `DefaultFilesystems`. I prefer to put the error definition of runtime test and bundle test together. Maybe I should also move the `validate/error.go` file to the `error` directory?